content/helpers.go

```diff
@@ -273,3 +330,58 @@ func copyWithBuffer(dst io.Writer, src io.Reader) (written int64, err error) {
 	}
 	return
 }
+
+func newTrylock() *trylock {
+	return &trylock{
+		locks: make(map[string]*lockItem),
+	}
+}
+
+type trylock struct {
+	sync.Mutex
+
+	locks map[string]*lockItem
+}
+
+type lockItem struct {
+	waiter *semaphore.Weighted
+	cnt    int
```

Comment:
Is `cnt` the number of writers who are waiting the lock?

Reply:
Yes. The unlock should check the number of waiter before deletion of the key.